rasa/nlu/extractors/crf_entity_extractor.py

```diff
@@ -130,216 +150,142 @@ def train(
+
+        Args:
+            predictions: list of mappings from entity tag to confidence value
+
+        Returns:
+            List of entity tags and list of confidence values.
+        """
+        _tags = []
+        _confidences = []
+
+        for token_predictions in predictions:
+            tag = max(token_predictions, key=lambda key: token_predictions[key])
+            _tags.append(tag)
 
-        else:
-            return None, None, None
+            if self.component_config["BILOU_flag"]:
+                # if we are using BILOU flags, we will sum up the prob
+                # of the B, I, L and U tags for an entity
+                _confidences.append(
+                    sum(
+                        _confidence
+                        for _tag, _confidence in token_predictions.items()
+                        if bilou_utils.tag_without_prefix(tag)
+                        == bilou_utils.tag_without_prefix(_tag)
+                    )
+                )
+            else:
+                _confidences.append(token_predictions[tag])
 
-    def _from_crf_to_json(
-        self, message: Message, entities: List[Any]
-    ) -> List[Dict[Text, Any]]:
+        return _tags, _confidences
 
-        tokens = self._tokens_without_cls(message)
+    def _tag_confidences(
+        self, tokens: List[Token], predictions: Dict[Text, List[Dict[Text, float]]]
+    ) -> Tuple[Dict[Text, List[Text]], Dict[Text, List[float]]]:
+        """Get most likely tag predictions with confidence values for tokens."""
+        tags = {}
+        confidences = {}
 
-        if len(tokens) != len(entities):
-            raise Exception(
-                "Inconsistency in amount of tokens between crfsuite and message"
-            )
+        for tag_name, predicted_tags in predictions.items():
+            if len(tokens) != len(predicted_tags):
+                raise Exception(
+                    "Inconsistency in amount of tokens between crfsuite and message"
+                )
 
-        if self.component_config["BILOU_flag"]:
-            return self._convert_bilou_tagging_to_entity_result(
-                message, tokens, entities
-            )
-        else:
-            # not using BILOU tagging scheme, multi-word entities are split.
-            return self._convert_simple_tagging_to_entity_result(tokens, entities)
+            _tags, _confidences = self._most_likely_tag(predicted_tags)
 
-    def _convert_bilou_tagging_to_entity_result(
-        self, message: Message, tokens: List[Token], entities: List[Dict[Text, float]]
-    ):
-        # using the BILOU tagging scheme
-        json_ents = []
-        word_idx = 0
-        while word_idx < len(tokens):
-            end_idx, confidence, entity_label = self._handle_bilou_label(
-                word_idx, entities
-            )
+            if self.component_config["BILOU_flag"]:
+                bilou_utils.ensure_consistent_bilou_tagging(_tags)
```

Comment:
doing this in DIET will be separate PR?

Reply:
I added a similar line to DIET in this PR.